### PR TITLE
Fixes #1590 - [FEATURE REQUEST] Ability to explicitly specify conversion to datetime or datetime2 

### LINKED ISSUE #1590: [FEATURE REQUEST] Ability to explicitly specify conversion to datetime or datetime2
## Is your feature request related to a problem? If so, please give a short summary of the problem and how the feature would resolve it
We have an application that uses SQL Server 2019 and the current schema uses `datetime` datatype for tracking timestamps. Since the database compatibility level is set to 2019, the JDBC driver is converting all timestamps into `datetime2` datatypes which is causing a number of bugs in our application.

We are currently using a third party JDBC driver (Progress® DataDirect Connect® Series for JDBC™) which offers a `DateTimeInputParameterType` connection string which you can use on newer databases to have it declare parameters as `datetime` instead of `datetimeoffset`. However, we are soon planning on moving our application stack and want to use the Microsoft SQL Server Driver as it is officially support in the new application server we plan on using.

I would like to see a similar connection string added to the JDBC driver which would allow you to specify how the JDBC driver handlers timestamps without having to lower the compatibility level of the database.

## Describe the preferred solution
Add a new connection string (e.g. `timestampEncoding`) which can be used to determine how timestamp should be encoded.  For example, you might have the following options:

Option | Description
------- | ------------
DATETIMEOFFSET | always use DATETIMEOFFSET.
DATETIME2 | always use DATETIME2. Good if the user knows they only use DATETIME2 datatypes in their schema (current behavior).
DATETIME | always use DATETIME (old behavior). Good if the user is dealing with a legacy schema and is not willing to upgrade.
QUERYDB | the new code, but not compatible with encrypted connections. Good for mixed schemas.

See https://github.com/microsoft/mssql-jdbc/issues/443#issuecomment-849720800

## Describe alternatives you've considered
Alternatives we've considered:

* Converting the schema to use `datetime2`. We will probably eventually do this, but in a large multi-tenant application stack this is going to take a lot of work. We have a lot of functionality driven by timestamps, so this refactoring will take time.
* Casting values to `datetime`. Timely to implement and would require another round of refactoring when we final do migrate to `datetime2` values.
* Lowering the database compatibility level. Unfortunately, we are using some SQL Server 2017 functionality, so in order to do this we would need to remove this functionality. We would also lose query plan performance improvements implemented in newer versions of SQL Server.
* Forking the current code and using our fork. We'd like to avoid forking the code if possible, so we can stay on the official code path.

## Additional context
none

## Reference Documentations/Specifications
n/a

### PR DESCRIPTION
**Pull request by dswitzer**
I've implemented a new connection string (i.e. `datetimeParameterType`) which fixes #1590 by allowing users to specify the datatype to use for date/timestamp parameters.

This pull request adds the ability to define the `datetimeParameterType` connection string with any of the following values:

Option | Description
------- | ------------
datetime2| always use DATETIME2. Good if the user knows they only use DATETIME2 datatypes in their schema (current behavior, this is the default).
datetime| always use DATETIME. Good if the user is dealing with a legacy schema and is not willing to upgrade.
datetimeoffset| always use DATETIMEOFFSET. 

By default, the value defaults to `datetime2`, which is the same behavior as today. Setting this value to any of the above will cause any SQL version of Katmai or higher to always use that datatype for date/time values when generating parameters. For older versions of SQL Server that do not support `datetime2` or `datetimeoffset`, then `datetime` is always used (just as it has been working).

I've added unit test coverage for the changes.

There should be no behavioral change for users who do not set the `datetimeParameterType` connection string value. It should work as it has been. However, when supporting older databases running in SQL Server 2016 or higher compatibility mode, the `datetimeParameterType` can be set to `datetime` to support databases only using that data type without worrying about casting issues in SQL Server with conversion between `datetime2` and `datetime`.

### CODE REVIEW

**Comment by v-zhangw** [comment]
Hello @dswitzer,
Could you please resolve the merge conflicts in this PR? Then we can run the pipeline tests to check for acceptance.

**Comment by dswitzer** [comment]
@v-zhangw ,

Ok, I believe the merge should be fixed, some of the checks are failing, but they not in portions of the code that should have been affected by my changes.

**Comment by dswitzer** [comment]
@v-zhangw

Just checking up to see if there's anything more you need from me.

**Comment by pacham7** [comment]
@lilgreenbird @v-zhangw, this pull request would be really useful, as updating to a later driver would unblock migration to jdk17 for us (ssl issues with existing very old driver). Is there any chance someone can take a look at it?

**Comment by dswitzer** [comment]
It should be pretty easy to resolve the conflicts, I just do not have write access to the repository to do it.

**Comment by AlBundy33** [comment]
@dswitzer not sure but I think you have to update your forked branch and fix the conflicts there.

**Comment by dswitzer** [comment]
@AlBundy33,

I'd be happy to update the fork, but the origin code keeps changing so fast, that the PR keeps becoming invalidated. It wasn't worth the time to keep fixing the code when there's no movement on them approving the PR. 

If they're willing to actually move on the PR, I'm happy to fix the conflicts.

I just don't want to keep spending time to fix my PR for no reason.

**Comment by AlBundy33** [comment]
@v-zhangw @lilgreenbird @Jeffery-Wasty what do you think about the pr?

**Comment by Jeffery-Wasty** [comment]
Hi @AlBundy33,

The issue and associated PR are on our radar and we'll look into both of them as soon as we are able to.

**Comment by lilgreenbird** [comment]
Hi @dswitzer 

Apologies this has been put off for so long. We had determined that the #1590 needed more investigation and then we have had problems juggling this as this a feature request so it kept getting bumped by other higher priority issues. We are now getting back to looking into this again. Once you get this PR up to date we will run this thru our test lab to verify.

**Comment by dswitzer** [comment]
I've updated the pull request. The failures do not appear to be related to my changes.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerLexer.java
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerLexer.java:23]: why is this change necessary?

Comment (dswitzer) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerLexer.java:23]: It shouldn't be. Mostly likely just a merge conflict. That's not something I purposefully changed.

File: src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java:628]: this check still need to be there for !isKatmaiOrLater()as DATATIME2 is not supported before

same with line 632

Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java:628]: perhaps we should log an error if it's set to datatime2 and pre 2008

**Review by lilgreenbird — COMMENTED** [review]
- please use our code formatter on all the changes.
- there seems to be an issue with the Mac CI pipeline line please ignore that
- fyi I'm in the process of running this in our test lab. As this adds a new API we will also need this to be approved by our product manager who's away this week, I'll ask him to review this next week.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java:217]: we don't usually add property value options in the error strings as there may be changed

File: src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java
Comment (lilgreenbird) [inline src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java:936]: I'm wondering how this should be handled, if one explicitly sets data2 but it's pre SQL Server 2008. Should this silently change to datetime or error or just log a warning? Don't make any changes here yet I'll have to check with our product manager next week to see which he prefers.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java
Comment (dswitzer) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java:217]: I was just trying to follow the pattern for what I saw in the code for the `sendTimeAsDatetime` option, as they seemed closely related.

**Comment by dswitzer** [comment]
@lilgreenbird,

> * please use our code formatter on all the changes.

When I run the _Format Document_ command from VS Code, it's making a lot of changes to the files in places of the code I did not change—especially in the comments. 

Is that what you're wanting me to do? 

If not, if you could provide me with explicit instructions on what you need me to do, I would appreciate it very much!

File: src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java
Comment (dswitzer) [reply src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java:936]: For pre-Katmai, I wouldn't do anything here. Maybe there should be a warning in the log if you try to set the value saying the parameter is unsupported.

For Katmai, I wasn't sure how you generally approach things like this. I was going to fault tolerance and trying to minimize any backwards compatibility issues. I think it would make some sense to throw a hard error if an invalid value is supplied, but a warning message certainly makes sense.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java
Comment (David-Engel) [inline src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:6917]: You can run all this through the `DatetimeType.valueOfString(datetimeParameterValueType)` method to validate the input parameter and eliminate the error message instead of hard coding more strings here.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerLexer.java
Comment (David-Engel) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerLexer.java:23]: This should be reverted.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java
Comment (David-Engel) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerResource.java:217]: It looks like we haven't been consistent in the past. I'm good with including the possible values as it's helpful to the user.

**Review by David-Engel — COMMENTED** [review]
I'm good with the public-facing aspects of this change (public APIs, connection string additions). I had one javadoc-related suggestion and I commented on a few code issues, too.

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerConnection.java:6917]: this will be updated in a subsequent PR

File: src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/Parameter.java:923]: this will be updated in a subsequent PR

**Review by lilgreenbird — APPROVED** [review]

**Review by tkyc — APPROVED** [review]

File: src/main/java/com/microsoft/sqlserver/jdbc/SQLServerLexer.java
Comment (lilgreenbird) [reply src/main/java/com/microsoft/sqlserver/jdbc/SQLServerLexer.java:23]:
```suggestion
        RuntimeMetaData.checkVersion("4.9.3", RuntimeMetaData.VERSION);

```

**Review by lilgreenbird — APPROVED** [review]

**Review by tkyc — APPROVED** [review]

**Comment by AlBundy33** [comment]
@lilgreenbird is this already available in a release?

**Comment by Jeffery-Wasty** [comment]
Hi @AlBundy33,

This is planned to be included in the 12.2.0 release, scheduled for the end of this month.

**Pull request merged**
